src/go/rpk/pkg/config/config.go

```diff
@@ -223,89 +145,52 @@ func AvailableModes() []string {
 	}
 }
 
-func Check(conf *Config) (bool, []error) {
-	configMap, err := toMap(conf)
-	if err != nil {
-		return false, []error{err}
-	}
-
-	v := viper.New()
-	err = v.MergeConfigMap(configMap)
-	if err != nil {
-		return false, []error{err}
-	}
-	return check(v)
-}
-
-func check(v *viper.Viper) (bool, []error) {
-	errs := checkRedpandaConfig(v)
+func Check(cfg *Config) (bool, []error) {
+	errs := checkRedpandaConfig(cfg)
 	errs = append(
 		errs,
-		checkRpkConfig(v)...,
+		checkRpkConfig(cfg)...,
 	)
 	ok := len(errs) == 0
 	return ok, errs
 }
 
-func checkRedpandaConfig(v *viper.Viper) []error {
-	errs := []error{}
-	if v.GetString("redpanda.data_directory") == "" {
+func checkRedpandaConfig(cfg *Config) []error {
+	var errs []error
+	rp := cfg.Redpanda
+	// top level check
+	if rp.Directory == "" {
 		errs = append(errs, fmt.Errorf("redpanda.data_directory can't be empty"))
 	}
-	if v.GetInt("redpanda.node_id") < 0 {
+	if rp.ID < 0 {
 		errs = append(errs, fmt.Errorf("redpanda.node_id can't be a negative integer"))
 	}
 
-	rpcServerKey := "redpanda.rpc_server"
-	exists := v.Sub(rpcServerKey) != nil
-	if !exists {
-		errs = append(
-			errs,
-			fmt.Errorf("%s missing", rpcServerKey),
-		)
+	// rpc server
+	if rp.RPCServer == (SocketAddress{}) {
+		errs = append(errs, fmt.Errorf("redpanda.rpc_server missing"))
 	} else {
-		socket := &SocketAddress{}
-		err := unmarshalKey(v, rpcServerKey, socket)
+		_, err := yaml.Marshal(rp.RPCServer)
 		if err != nil {
-			errs = append(
-				errs,
-				fmt.Errorf("invalid structure for %s", rpcServerKey),
-			)
+			errs = append(errs, fmt.Errorf("invalid structure for redpanda.rpc_server"))
 		} else {
 			errs = append(
 				errs,
-				checkSocketAddress(*socket, rpcServerKey)...,
+				checkSocketAddress(rp.RPCServer, "redpanda.rpc_server")...,
 			)
 		}
 	}
 
-	kafkaAPIKey := "redpanda.kafka_api"
-	exists = v.Get(kafkaAPIKey) != nil
-	if !exists {
-		errs = append(
-			errs,
-			fmt.Errorf("%s missing", kafkaAPIKey),
-		)
+	// kafka api
+	if len(rp.KafkaAPI) == 0 {
+		errs = append(errs, fmt.Errorf("redpanda.kafka_api missing"))
 	} else {
-		var kafkaListeners []NamedSocketAddress
-		err := unmarshalKey(v, "redpanda.kafka_api", &kafkaListeners)
+		_, err := yaml.Marshal(rp.KafkaAPI)
```

Comment:
How can this marshal fail? Isn't `rp.KafkaAPI` populated by an _unmarshal_?

Reply:
It can't. That's an unnecessary check. :+1: